library/core/src/iter/adapters/zip.rs

```diff
@@ -434,7 +434,7 @@ impl<A: Debug + TrustedRandomAccess, B: Debug + TrustedRandomAccess> ZipFmt<A, B
 ///    called on `self`:
 ///     * `std::clone::Clone::clone()`
 ///     * `std::iter::Iterator::size_hint()`
-///     * `std::iter::Iterator::next_back()`
+///     * `std::iter::DoubleEndedIterator::next_back()`
```

Comment:
I think this is neither correct nor used this way by current implementations. E.g. zip specializes next_back using `__iterator_get_unchecked` too which means calling `next_back()` on zip after using `__iterator_get_unchecked` on zip will yield wrong results.

Reply:
So you would suggest removing the line entirely?

Reply:
```
    #[inline]
    fn next_back(&mut self) -> Option<(A::Item, B::Item)>
    where
        A: DoubleEndedIterator + ExactSizeIterator,
        B: DoubleEndedIterator + ExactSizeIterator,
    {
        if A::MAY_HAVE_SIDE_EFFECT || B::MAY_HAVE_SIDE_EFFECT {
            let sz_a = self.a.size();
            let sz_b = self.b.size();
            // Adjust a, b to equal length, make sure that only the first call
            // of `next_back` does this, otherwise we will break the restriction
            // on calls to `self.next_back()` after calling `get_unchecked()`.
            if sz_a != sz_b {
                let sz_a = self.a.size();
                if A::MAY_HAVE_SIDE_EFFECT && sz_a > self.len {
                    for _ in 0..sz_a - self.len {
                        self.a.next_back();
                    }
                    self.a_len = self.len;
                }
                let sz_b = self.b.size();
                if B::MAY_HAVE_SIDE_EFFECT && sz_b > self.len {
                    for _ in 0..sz_b - self.len {
                        self.b.next_back();
                    }
                }
            }
        }
        if self.index < self.len {
            self.len -= 1;
            self.a_len -= 1;
            let i = self.len;
            // SAFETY: `i` is smaller than the previous value of `self.len`,
            // which is also smaller than or equal to `self.a.len()` and `self.b.len()`
            unsafe {
                Some((self.a.__iterator_get_unchecked(i), self.b.__iterator_get_unchecked(i)))
            }
        } else {
            None
        }
    }
```
to me looks like `.next_back()` _is_ being used. And `next` calls `__iterator_get_unchecked`. So the underlying iterator is queried by `next_back()` following an `__iterator_get_unchecked` operation if you call `next()` followed by `next_back()` on the `Zip`

Reply:
> me looks like .next_back() is being used.

Yeah, I guess the condition is more complicated. If an outer wrapper implements `next_back()` as `inner.__iterator_get_unchecked(inner.len() - offset)` then it can't switch back to using `next_back()` on the inner. Zip can't either for that matter, it only uses it when `next_back()` called for the first time.

Things would start to get really dicy once some specialization made use of the *Random* in `TrustedRandomAccess`, currently they're all just taking from the front or the back.